ksql-common/src/main/java/io/confluent/ksql/function/UdfIndex.java

```diff
@@ -0,0 +1,317 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.function;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
+import io.confluent.ksql.function.udf.UdfMetadata;
+import io.confluent.ksql.util.KsqlException;
+import io.confluent.ksql.util.SchemaUtil;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.function.BiPredicate;
+import java.util.stream.Collectors;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.Schema.Type;
+
+/**
+ * An index of all method signatures associated with a UDF. This index
+ * supports lookups based on method signatures, defined by a list of
+ * nullable {@link Schema}.
+ *
+ * <p>Resolving a method signature takes the following precedence rules:
+ * <ul>
+ *   <li>If only one exact method exists, return it</li>
+ *   <li>If a method exists such that all non-null {@code Schema} in the
```

Comment:
Question about the null argument handling... an argument would only be null if the user explicitly passed a `NULL` literal to the UDF, right?  Otherwise we know the type of a column, even if a row has a `null` value for a specific column, right?

Reply:
I believe you are correct - we get the arguments from the DML, not at runtime.